### ISSUE TITLE
bughuntoor - `updateUserDeposit` overwrites current values

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-03-zap-protocol-judging#55]
bughuntoor

medium

# `updateUserDeposit` overwrites current values

## Summary
`updateUserDeposit` overrides current values 

## Vulnerability Detail
If we look at `updateUserDeposit`, we'll see that the amount to be vested to the user will get overwritten, instead of increased. 
```solidity
    function updateUserDeposit(
        address[] memory _users,
        uint256[] memory _amount
    ) public onlyRole(DEFAULT_ADMIN_ROLE) {
        require(_users.length <= 250, "array length should be less than 250");
        require(_users.length == _amount.length, "array length should match");
        uint256 amount;
        for (uint256 i = 0; i < _users.length; i++) {
            userdetails[_users[i]].userDeposit = _amount[i];
            amount += _amount[i];
        }
        token.safeTransferFrom(distributionWallet, address(this), amount);
    }
```
Despite, the idea of the function being to update the existing value, If the user has prior balance, it will be lost and the tokens will be forever stuck. 

## Impact
Loss of funds

## Code Snippet
https://github.com/sherlock-audit/2024-03-zap-protocol/blob/main/zap-contracts-labs/contracts/Vesting.sol#L61

## Tool used

Manual Review

## Recommendation
Increase the current values instead of overwriting them

**Comment by ZdravkoHr** [comment]
Escalate

This function is not used anywhere in the whole project, it's intended to be called by the `DEFAULT_ADMIN`. Overwriting the previous value instead of increasing it is a design choice. Deleting the old value is basically an admin input mistake which is invalid according to the Sherlock rules because admins are `TRUSTED`.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> This function is not used anywhere in the whole project, it's intended to be called by the `DEFAULT_ADMIN`. Overwriting the previous value instead of increasing it is a design choice. Deleting the old value is basically an admin input mistake which is invalid according to the Sherlock rules because admins are `TRUSTED`.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by koreanspicygarlic1** [comment]
This could not be considered a design mistake - overwriting the values results in permanently stuck funds within the contract.

**Comment by detectiveking123** [comment]
Agree with escalation. This is a design decision. The admin would have to make a mistake to overwrite the values.

Also there is no permanently stuck funds inside the contract, even if the admin magically makes a mistake. Admin can just re-update to the correct values and then extract out the excess ETH at the end (`takeLockedBNB` function).

**Comment by mosharaf6** [comment]
The issue is not valid !
As my understanding , the function will be called once for a user after the tokensale has ended . There'e  no chance of overriding in that case . If any problem arises from this function , that would only be a admin mistake .

**Comment by vsharma4394** [comment]
I had asked from the sponsers whether this function will be called more than once for a user and they said no, it will only be called once for a user which makes this finding invalid.

**Comment by MatinR1** [comment]
The issue is valid!

If we look in detail at the docs of Sherlock, we can see this:

> When external-admin=trusted, issues related to these external admins being able to rug protocol users is not a valid issue. (Example: Aave governance has the intention of rugging Index Protocol)

and also [Admin Input/call validation section](https://docs.sherlock.xyz/audits/judging/judging#vii.-list-of-issue-categories-that-are-not-considered-valid),

we can realize that issues related to admin mistakes or rug vectors are out of scope.
However, this issue is not a simple admin input mistake nor a rug pull case from the user assets. This issue impacts the uncontrolled drainage of the distribution wallet which in some cases may affect the system in an undesirable manner.
I clearly demonstrated this issue in a POC in issue [#17 ]

**Comment by detectiveking123** [comment]
Your point doesn't make sense. Adding some amount `amount` to a variable `a` is the same as setting it to `a + amount`. The current function allows the admin even more functionality than simply incrementing amounts because it also allows them to decrement.

**Comment by ghost** [comment]
It is borderline low/Medium, as admin input is low according to sherlock's docs, juxtaposed with the potential severe impact in the event of incorrect input, presents a nuanced perspective on risk assessment. I would let @Czar102 decide this one.

**Comment by shubham-antier** [comment]
Fixed the overwriting issue.

**Comment by sherlock-admin4** [comment]
The protocol team fixed this issue in the following PRs/commits:
https://github.com/Lithium-Ventures/zap-contracts-labs/pull/5

**Comment by vsharma4394** [comment]
I hadn't reported this issue because i had clearly asked the sponsers whether the value of userdetails[_users[i]].userDeposit would ever be changed/increased/decreased  and they said no, and this will be called only once for a user so no point of increasing/decreasing its value,thus making this issue invalid.

**Comment by ghost** [comment]
Hey I can't comment on your conversation with sponsors as I am not aware of it but mostly this would be a low severity issue

**Comment by Evert0x** [comment]
Planning to accept escalation and invalidate issue.

As protocol admin = trusted, this change will serve as a sanity check to make sure the admins don't overwrite the current values.

In the current implementation it's already possible to increase the current values.

> Admin Input/call validation: Protocol admin is considered to be trusted in most cases, hence issues where
> 1. Admin incorrectly enters an input parameter. Example: Make sure interestPerMin > 1 ether as it is an important parameter. This is not a valid issue.
>

**Comment by Evert0x** [comment]
Result:
Invalid  
Has Duplicates

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [ZdravkoHr](https://github.com/sherlock-audit/2024-03-zap-protocol-judging/issues/55/#issuecomment-2025373747): accepted

**Comment by sherlock-admin2** [comment]
The Lead Senior Watson signed off on the fix.
